pkg/registry/core/service/allocator/patch_bitmap.go

```diff
@@ -0,0 +1,73 @@
+/*
+Copyright 2015 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package allocator
+
+import (
+	"math/big"
+	"math/rand"
+	"time"
+)
+
+// NewAllocationMapReserved creates an allocation bitmap using a modified random scan strategy
+// that maintains a reserved offset for OpenShift.
+func NewAllocationMapReserved(max int, rangeSpec string) *AllocationBitmap {
+	// OpenShift Reserved Offsets:
+	reserved := make(map[int]struct{})
+	// - OpenShift DNS always uses the .10 address (0 counts so we reserve the 9 offset)
+	reserved[9] = struct{}{}
+
+	a := AllocationBitmap{
+		strategy: randomScanStrategyReserved{
+			rand:     rand.New(rand.NewSource(time.Now().UnixNano())),
+			reserved: reserved,
+		},
+		allocated: big.NewInt(0),
+		count:     0,
+		max:       max,
+		rangeSpec: rangeSpec,
+	}
+	return &a
+}
+
+// randomScanStrategyReserved chooses a random address from the provided big.Int, and then
+// scans forward looking for the next available address (it will wrap the range if necessary).
+// randomScanStrategyReserved omit some offsets that are "special" so they can't be allocated
+// randomly, only explicitly
+type randomScanStrategyReserved struct {
+	rand     *rand.Rand
+	reserved map[int]struct{}
+}
+
+func (rss randomScanStrategyReserved) AllocateBit(allocated *big.Int, max, count int) (int, bool) {
+	if count >= max {
```

Comment:
Should you account for the reserved address here, and check if `count + 1 >= max`?  E.g. if you have assigned everything except `.10`, the reserved one, yet something is still trying to allocate an address with this, will it keep trying even though all addresses have been either handed out or reserved?

Reply:
the number of addresses available remains the same, the only difference is that the reserved can not be allocated randomly, but still can be allocated directly and should be counted.
The for loop below skip the reserved values and exit after iterating over the whole range

Reply:
Let me put it another way - when all that is left is the reserved address, I expect that this will just return `0, false` without iterating over the whole range. Does it do that?

Reply:
that is indeed an optimization, but the problem is that you don't know beforehand if the missing address is the reserved or not.
Assume an allocator with size 4 and the value 1 is the reserved.
Case A:
(1,2,3) values reserved, count=3 , we bail out and don't allocate 4
Case B:
(2,3,4) values reserved, count=3, we bail out because we can't allocate 1 here that is correct

The algorithms keeps being O(n) and the code is much simpler, I think we are good this way

Reply:
added `TestPostAllocateReservedFull_BitmapReserved` and `TestPreAllocateReservedFull_BitmapReserved` test cases